packages/zoe/src/zoeService/purseStorage.js

```diff
@@ -0,0 +1,103 @@
+// @ts-check
+
+import { amountMath } from '@agoric/ertp';
+import { E } from '@agoric/eventual-send';
+import { makeWeakStore as makeNonVOWeakStore } from '@agoric/store';
+
+import { arrayToObj, objectMap } from '../objArrayConversion';
+
+export const makePurseStorage = () => {
+  /** @type {WeakStore<Brand, ERef<Purse>>} */
+  const brandToPurse = makeNonVOWeakStore('brand');
+
+  /**
+   * Create a purse for a new issuer
+   *
+   * @param {IssuerRecord} record
+   * @returns {void}
+   */
+  const createPurse = record => {
+    if (!brandToPurse.has(record.brand)) {
+      brandToPurse.init(record.brand, E(record.issuer).makeEmptyPurse());
+    }
+  };
+
+  /**
+   * Create a purse for a new, local issuer. Used only for ZCFMint issuers.
+   *
+   * @param {IssuerRecord} record
+   * @returns {Purse}
+   */
+  const createLocalPurse = record => {
+    const localPurse = record.issuer.makeEmptyPurse();
+    brandToPurse.init(record.brand, localPurse);
+    return localPurse;
+  };
+
+  /** @type {WithdrawPayments} */
+  const withdrawPayments = allocation => {
+    return harden(
+      objectMap(allocation, ([keyword, amount]) => {
+        const purse = brandToPurse.get(amount.brand);
+        return [keyword, E(purse).withdraw(amount)];
+      }),
+    );
+  };
+
+  /**
+   *
+   *  Only used internally. Actually deposit a payment or promise for payment.
+   *
+   * @param {ERef<Payment>} paymentP
+   * @param {Amount} amount
+   * @returns {Promise<Amount>}
+   */
+  const doDepositPayments = (paymentP, amount) => {
+    const purse = brandToPurse.get(amount.brand);
+    return E.when(paymentP, payment => E(purse).deposit(payment, amount));
+  };
+
+  // Proposal is cleaned, but payments are not
+
+  /**
+   * Deposits payments or promises for payments according to the
+   * `give` property of the proposal. Using the proposal, creates an
+   * initial allocation including the amount deposited for `give`
+   * keywords and an empty amount for `want` keywords.
+   *
+   * @param {ProposalRecord} proposal
+   * @param {PaymentPKeywordRecord} payments
+   * @returns {Promise<Allocation>}
+   */
+  const depositPayments = async (proposal, payments) => {
+    const { give, want } = proposal;
+    const giveKeywords = Object.keys(give);
+    const wantKeywords = Object.keys(want);
+    const proposalKeywords = harden([...giveKeywords, ...wantKeywords]);
+
+    const amountsDeposited = await Promise.all(
```

Comment:
Would be good to mention here the important problem we're not yet trying or claiming to solve: If any of these deposits hang or fail, then `depositPayments` hangs of fails, the offer does not succeed, and any funds that were deposited into the pool purses are lost.

(We should still get back to this someday. But it will require significant mechanism, our system is coherent without it, and the offer safety we still provide is still crisp and meaningful. Perhaps add to the comment a link to the issue.)

Reply:
Good idea, I'll link to the issue.